types/lib/stdlib.go

```diff
@@ -0,0 +1,269 @@
+	}
+	store := ctx.KVStore(lm.key)
+	bz, err := lm.cdc.MarshalBinary(value)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(lm.ElemKey(index), bz)
+}
+
+func (lm listMapper) Delete(ctx sdk.Context, index int64) {
+	if index < 0 {
+		panic(fmt.Errorf("Invalid index in ListMapper.Delete(ctx, %d)", index))
+	}
+	store := ctx.KVStore(lm.key)
+	store.Delete(lm.ElemKey(index))
+}
+
+func (lm listMapper) Push(ctx sdk.Context, value interface{}) {
+	length := lm.Len(ctx)
+	lm.Set(ctx, length, value)
+
+	store := ctx.KVStore(lm.key)
+	store.Set(lm.LengthKey(), marshalInt64(lm.cdc, length+1))
+}
+
+func (lm listMapper) Iterate(ctx sdk.Context, ptr interface{}, fn func(sdk.Context, int64) bool) {
+	length := lm.Len(ctx)
+	for i := int64(0); i < length; i++ {
+		lm.Get(ctx, i, ptr)
+		if fn(ctx, i) {
+			break
+		}
+	}
+}
+
+func (lm listMapper) LengthKey() []byte {
+	return []byte(fmt.Sprintf("%s/%d", lm.prefix, lm.lk))
+}
+
+func (lm listMapper) ElemKey(i int64) []byte {
+	return []byte(fmt.Sprintf("%s/%d", lm.prefix, i))
+}
+
+// QueueMapper is a Mapper interface that provides queue-like functions
+// It panics when the element type cannot be (un/)marshalled by the codec
+
+type QueueMapper interface {
+	Push(sdk.Context, interface{})
+	// Popping/Peeking on an empty queue will cause panic
+	// The user should check IsEmpty() before doing any actions
+	Peek(sdk.Context, interface{})
+	Pop(sdk.Context)
+	IsEmpty(sdk.Context) bool
+	// Iterate() removes elements it processed; return true in the continuation to break
+	Iterate(sdk.Context, interface{}, func(sdk.Context) bool)
+	Info(sdk.Context) QueueInfo
+}
+
+type queueMapper struct {
+	key    sdk.StoreKey
+	cdc    *wire.Codec
+	prefix string
+	lm     ListMapper
+	lk     []byte
+	ik     []byte
+}
+
+func NewQueueMapper(cdc *wire.Codec, key sdk.StoreKey, prefix string) QueueMapper {
+	lk := []byte("list")
+	ik := []byte("info")
+	return queueMapper{
+		key:    key,
+		cdc:    cdc,
+		prefix: prefix,
+		lm:     NewListMapper(cdc, key, prefix+string(lk)),
+		lk:     lk,
+		ik:     ik,
+	}
+}
+
+type QueueInfo struct {
+	// begin <= elems < end
+	Begin int64
+	End   int64
+}
+
+func (info QueueInfo) validateBasic() error {
+	if info.End < info.Begin || info.Begin < 0 || info.End < 0 {
+		return fmt.Errorf("Invalid queue information: {Begin: %d, End: %d}", info.Begin, info.End)
+	}
+	return nil
+}
+
+func (info QueueInfo) isEmpty() bool {
+	return info.Begin == info.End
+}
+
+func (qm queueMapper) getQueueInfo(store sdk.KVStore) QueueInfo {
+	bz := store.Get(qm.InfoKey())
+	if bz == nil {
+		store.Set(qm.InfoKey(), marshalQueueInfo(qm.cdc, QueueInfo{0, 0}))
+		return QueueInfo{0, 0}
+	}
+	var info QueueInfo
+	if err := qm.cdc.UnmarshalBinary(bz, &info); err != nil {
+		panic(err)
+	}
+	if err := info.validateBasic(); err != nil {
+		panic(err)
+	}
+	return info
+}
+
+func (qm queueMapper) setQueueInfo(store sdk.KVStore, info QueueInfo) {
```

Comment:
Why are `Begin` / `End` set and read together? Most operations only need to read/write one or the other.